[Beta]: docs(cn): extracting-state-logic-into-a-reducer (#693)
reactjs/zh-hans.reactjs.org: beta/src/pages/learn/extracting-state-logic-into-a-reducer.md

---

>  title: 把状态逻辑移到 Reducer 中

Reviewer: ```suggestion
title: 把状态逻辑移到 Reducer 中
```

translators:

>   - qinhua

Reviewer: ```suggestion
  - qinhua
  - yyyang1996
```

---

<Intro>

对于拥有许多状态更新逻辑的组件来说，过于分散的事件处理程序可能会令人不知所措。对于这种情况，你可以将组件的所有状态更新逻辑整合到一个外部函数中，这个函数叫作 “reducer”。

</Intro>

<YouWillLearn>

- 什么是 reducer 函数
- 如何将 `useState` 重构成 `useReducer`
- 什么时候使用 reducer
- 如何编写一个好的 reducer

</YouWillLearn>

## 使用 reducer 整合状态逻辑 {/*consolidate-state-logic-with-a-reducer*/}

> 随着组件复杂度的增加，你将很难一眼看清所有组件的状态更新逻辑。例如，下面的 `TaskBoard` 组件有一个数组类型的状态 `tasks`，并通过三个不同的事件处理程序来实现添加、删除和修改：

Reviewer: ```suggestion
随着组件复杂度的增加，你将很难一眼看清所有的组件状态更新逻辑。例如，下面的 `TaskBoard` 组件有一个数组类型的状态 `tasks`，并通过三个不同的事件处理程序来实现任务的添加、删除和修改：
```

</Sandpack>

这个组件的每个事件处理程序都通过 `setTasks` 来更新状态。随着这个组件的不断迭代，其状态逻辑也会越来越多。为了降低复杂度，且让所有逻辑都可以存放在一个易于理解的地方，你可以将这些状态逻辑移到组件之外的一个称为 **reducer** 的函数中。

> Reducer 是处理状态的另一种方式。你可以通过三个步骤从 `useState` 迁移到 `useReducer`：

Reviewer: ```suggestion
Reducer 是处理状态的另一种方式。你可以通过三个步骤将 `useState` 迁移到 `useReducer`：
```

</Sandpack>

> 这个组件的每个事件处理程序都通过 `setTasks` 来更新状态。随着这个组件的不断迭代，其状态逻辑也会越来越多。为了降低复杂度，且让所有逻辑都可以存放在一个易于理解的地方，你可以将这些状态逻辑移到组件之外的一个称为 **reducer** 的函数中。

Reviewer 2: ```suggestion
这个组件的每个事件处理程序都通过 `setTasks` 来更新状态。随着这个组件的不断迭代，其状态逻辑也会越来越多。为了降低这种复杂度，并让所有逻辑都可以存放在一个易于理解的地方，你可以将这些状态逻辑移到组件之外的一个称为 **reducer** 的函数中。
```

</Sandpack>

分离关注点可以让我们更容易地理解组件逻辑。现在事件处理程序只通过派发 `actions` 来指定 *发生了什么*，而 `reducer` 函数决定了 *状态如何更新* 来响应它们。

## `useState` 和 `useReducer` 的对比 {/*comparing-usestate-and-usereducer*/}

Reducers 并非没有缺点！你可以从这些角度去对比它们：

- **代码体积：** 通常，在使用 `useState` 时，一开始只需要编写少量代码。而 `useReducer` 必须提前编写 reducer 函数和需要调度的 actions。但是，当多个事件处理程序以相似的方式修改 state 时，`useReducer` 可以减少代码量。
- **可读性：** 当状态更新逻辑足够简单时，`useState` 的可读性还行。但是，一旦逻辑变得复杂起来，它们会使组件变得臃肿且难以扫描。在这种情况下，`useReducer` 允许你将状态更新逻辑与事件处理程序分离开来。
- **可调试性：** 当使用 `useState` 出现问题时, 你很难发现具体原因以及为什么。 而使用 `useReducer` 时， 你可以在 reducer 函数中通过打印日志的方式来观察每个状态的更新，以及为什么要更新 (来自哪个 `action`)。 如果所有 `action` 都没问题，你就知道问题出在了 reducer 本身的逻辑中。 然而，与使用 `useState` 相比，你必须遍历更多的代码。
- **可测试性：** reducer 是一个不依赖于组件的纯函数。这就意味着你可以单独对它进行测试。一般来说，我们最好是在真实环境中测试组件，但对于复杂的状态更新逻辑，把 reducer 断言为特定的初始状态和 `action` 可能会很有帮助。
- **个人偏好：** 并不是所有人都喜欢用 reducer，没关系，这是个人偏好问题。你可以随时在 `useState` 和 `useReducer` 之间切换，他们是等价的！

如果你在修改某些组件状态时经常出现问题或者想给组件添加更多逻辑时，我们建议你还是使用 reducer。当然，你也不必整个项目都用 reducer，这是可以自由搭配的。你甚至可以在一个组件中同时使用 `useState` 和 `useReducer`。

## 编写一个好的 reducers {/*writing-reducers-well*/}

编写 `reducers` 时最好牢记以下两点：

> - **reducers 必须是纯净的。** 这一点和 [状态更新函数](/learn/queueing-a-series-of-state-updates) 是相似的，`reducers` 在是在渲染时运行的！（actions 会排队直到下一次渲染)。 这就意味着 `reducers` [必须纯净](/learn/keeping-components-pure)，即当输入一定时，输出也是一定的。 (operations that impact things outside the component).它们不应该包含异步请求、定时器或者任何副作用（对组件外部有影响的操作）。它们应该以不可变值的方式去修改 [对象](/learn/updating-objects-in-state) 和 [数组](/learn/updating-arrays-in-state)。

Reviewer: ```suggestion
- **reducers 必须是纯净的。** 这一点和 [状态更新函数](/learn/queueing-a-series-of-state-updates) 是相似的，`reducers` 在是在渲染时运行的！（actions 会排队直到下一次渲染)。 这就意味着 `reducers` [必须纯净](/learn/keeping-components-pure)，即当输入相同时，输出也是相同的。它们不应该包含异步请求、定时器或者任何副作用（对组件外部有影响的操作）。它们应该以不可变值的方式去更新 [对象](/learn/updating-objects-in-state) 和 [数组](/learn/updating-arrays-in-state)。
```

</Sandpack>

> Reducers 应该是纯净的，所以它们不应该去修改 state。而 Immer 为你提供了一种特殊的 `draft` 对象，你可以用它安全的修改 state。在底层，Immer 会基于当前 state 创建一个副本。这就是为什么通过 `useImmerReducer` 来管理 reducers 时可以修改第一个参数，但不需要返回一个新的 state 的原因。

Reviewer: ```suggestion
Reducers 应该是纯净的，所以它们不应该去修改 state。而 Immer 为你提供了一种特殊的 `draft` 对象，你可以通过它安全的修改 state。在底层，Immer 会基于当前 state 创建一个副本。这就是为什么通过 `useImmerReducer` 来管理 reducers 时，可以修改第一个参数，且不需要返回一个新的 state 的原因。
```

<Recap>

* 把 `useState` 转化为 `useReducer`：

>   1. 通过事件处理函数派发 actions；

Reviewer: ```suggestion
  1. 通过事件处理函数 dispatch actions；
```

  2. 编写一个 reducer 函数，它接受传入的 state 和一个 action，并返回一个新的 state；
  3. 使用 `useReducer` 替换 `useState`；
* Reducers 可能需要你写更多的代码，但是这有利于代码的调试和测试。
* Reducers 必须是纯净的。
* Actions 描述的是 “发生了什么” 而不是 “要做什么”。

> * 使用 Immer 来帮你编写变更风格的 reducers。

Reviewer: ```suggestion
* 使用 Immer 来帮助你在 reducer 里直接修改状态。
```

</Recap>

<Challenges>

> ### 通过事件处理函数派发 actions {/*dispatch-actions-from-event-handlers*/}

Reviewer: ```suggestion
### 通过事件处理函数 dispatch actions {/*dispatch-actions-from-event-handlers*/}
```

目前，`ContactList.js` 和 `Chat.js` 中的事件处理程序包含 `// TODO` 注释。这就是为什么输入不起作用，点击按钮也不会改变收件人的原因。

> 将这两个 `// TODO` 替换为 `dispatch` 相应的操作。如果要查看 actions 对于的参数和类型，请阅读 `messerreducer.js` 中的 reducer。reducer 已经写好了，不需要修改它。你只需要在 `ContactList.js` 和 `Chat.js` 中派发动作即可。

Reviewer: ```suggestion
将这两个 `// TODO` 替换为 `dispatch` 相应的 action。如果要查看 action 的结构和类型，请查看 `messerreducer.js` 中的 reducer。reducer 已经写好了，你不需要再修改它。你只需要在 `ContactList.js` 和 `Chat.js` 中 dispatch 相应的 action 即可。
```

<Hint>

> `dispatch` 函数在这两个组件中都是可用的，因为它已经以 prop 的形式传递进来了。所以你需要用相应的 action 对象来调用 `dispatch` 函数。

Reviewer: ```suggestion
`dispatch` 函数在这两个组件中都是可用的，因为它已经以 prop 的形式传递进来了。因此你需要通过传入相应的 action 对象来调用 `dispatch` 函数。
```

}
```

> 这表示你的 `action` 对象应该有一个 `type: 'changed_selection'`。同时你也可以看到 `action.contactId` 被使用到了，所以你需要传入一个 `contactId` 属性到你的动作中。

Reviewer: ```suggestion
这表示你的 `action` 对象应该有一个 `type: 'changed_selection'`。同时你也可以看到 `action.contactId` 被使用到了，所以你需要传入一个 `contactId` 属性到你的 action 对象中。
```

<Solution>

> 通过 reducer 的代码，你可以推断出 actions 需要像下面这样：

Reviewer: ```suggestion
从 reducer 函数的代码中，你可以推断出 actions 需要像下面这样：
```

<Solution>

> 在 “发送” 按钮的事件处理程序中，有很多方法可以用来清空输入框。一种方法是显示一个 alert，然后派发一个带有空 `message` 的 `edited_message` 操作：

Reviewer: ```suggestion
在 “发送” 按钮的事件处理程序中，有很多方法可以用来清空输入框。一种方法是显示一个 alert，然后 dispatch 一个名为 `edited_message` 且带有空 `message` 的 action：
```

</Sandpack>

这样当你点击 “发送” 按钮时就会清空输入框。

> 然而，从用户的角度来看，发送消息与编辑字段是不同的操作。为了体现这一点，你可以创建一个名为 `sent_message` 的 *新动作*，并在 reducer 中单独处理：

Reviewer: ```suggestion
然而，从用户的角度来看，发送消息与编辑字段是不同的操作。为了体现这一点，你可以创建一个名为 `sent_message` 的新 *action*，并在 reducer 中单独处理：
```

</Sandpack>

结果虽然是一样的。但请记住，action 的类型应该准确描述 “用户做了什么”，而不是 “你希望状态如何改变”。这使得以后添加更多特性变的容易。

> 在任何一种解决方案中，最重要的是你 **不要** 把 `alert` 放置在 reducer 中。reducer 必须是一个纯函数——它应该只计算下一个状态。而不应该 “做” 其它事情，包括向用户显示消息。这应该在事件处理程序中处理。（为了便于捕获这样的错误，React 会在严格模式下多次调用你的 reducer。这就是为什么当你在 reducer 中加入一个 alert，它会触发两次的原因。）

Reviewer: ```suggestion
不管是哪一种解决方案，最重要的是你 **不要** 把 `alert` 放置在 reducer 中。reducer 必须是一个纯函数——它应该只计算下一个状态。而不应该 “做” 其它事情，包括向用户显示消息。这应该在事件处理程序中处理。（为了便于捕获这样的错误，React 会在严格模式下多次调用你的 reducer。这就是为什么当你在 reducer 中加入一个 alert，它会触发两次的原因。）
```

</Solution>

### 切换 tab 时重置输入框内容 {/*restore-input-values-when-switching-between-tabs*/}

> In this example, switching between different recipients always clears the text input:

Reviewer: 在这个示例中，切换收件人时总是会清空输入框。

</Sandpack>

结果虽然是一样的。但请记住，action 的类型应该准确描述 “用户做了什么”，而不是 “你希望状态如何改变”。这使得以后添加更多特性变的容易。

不管是哪一种解决方案，最重要的是你 **不要** 把 `alert` 放置在 reducer 中。reducer 必须是一个纯函数——它应该只计算下一个状态。而不应该 “做” 其它事情，包括向用户显示消息。这应该在事件处理程序中处理。（为了便于捕获这样的错误，React 会在严格模式下多次调用你的 reducer。这就是为什么当你在 reducer 中加入一个 alert，它会触发两次的原因。）

</Solution>

### 切换 tab 时重置输入框内容 {/*restore-input-values-when-switching-between-tabs*/}

在这个示例中，切换收件人时总是会清空输入框。

```js
case 'changed_selection': {
  return {
    ...state,
    selectedId: action.contactId,
    message: '' // 清空输入框
  };
```

这是因为你不希望在多个收件人之间共享单个邮件草稿。但如果你的应用程序能单独 “记住” 每个联系人的草稿，并在你切换联系人时恢复，那就更好了。

> 你的任务是改变状态的组织形式，以便能记住 _每个联系人_ 的消息草稿。你需要对 reducer、初始状态和组件进行一些修改。

Reviewer: ```suggestion
你的任务是改变状态的组织形式，以便能记住 *每个联系人* 的消息草稿。你需要对 reducer、初始状态和组件进行一些修改。
```

</Sandpack>

> 显然，为了实现不同行为，你不需要修改任何事件处理程序。但如果没使用 reducer 的话，你不得不在每个事件处理程序中去更新状态。

Reviewer: ```suggestion
显然，你不再需要通过修改任何事件处理程序来实现不同的行为。但如果没使用 reducer 的话，你不得不在每个事件处理程序中去更新状态。
```

</Solution>

### 从零开始实现 `useReducer` {/*implement-usereducer-from-scratch*/}

> 在前面的例子中，你从 React 中导入了 `useReducer` 钩子。这一次，你将自己实现 `useReducer` 钩子。这里有一个模板，你可以从它开始。它不应该超过 10 行代码。

Reviewer: ```suggestion
在前面的例子中，你从 React 中导入了 `useReducer` Hook。现在，你将学习自己实现 `useReducer` Hook。你可以从这个模板开始，它不会超过 10 行代码。
```

      />
      <br />
      <button onClick={() => {
        alert(`正在发送 "${message}" 到 ${contact.email}`);
        dispatch({

>            type: 'sent_message',

Reviewer: ```suggestion
          type: 'sent_message',
```

<Solution>

> 派发一个 action 去调用一个具有当前 state 和 action 的 reducer，并将结果存储为下一个 state。下面是它在代码中的样子：

Reviewer: ```suggestion
dispatch 一个 action 去调用一个具有当前 state 和 action 的 reducer，并将结果存储为下一个 state。下面是它在代码中的样子：
```

</Sandpack>

虽然在大多数情况下这并不重要，但更准确的实现是这样的：

```js
function dispatch(action) {
  setState(s => reducer(s, action));
}
```

> 这是因为被派发的 actions 在下一次渲染之前都是处于排队状态的，[和更新函数类似](/learn/queueing-a-series-of-state-updates)。

Reviewer: ```suggestion
这是因为被派发的 actions 在下一次渲染之前都是处于排队状态的，这和 [状态更新函数](/learn/queueing-a-series-of-state-updates) 类似。
```
